lib/usd/pxrUsdPreviewSurface/usdPreviewSurfaceReader.cpp

```diff
@@ -0,0 +1,160 @@
+    PxrMayaUsdPreviewSurface_Reader(const UsdMayaPrimReaderArgs&);
+
+    bool Read(UsdMayaPrimReaderContext* context) override;
+
+    TfToken GetMayaNameForUsdAttrName(const TfToken& usdAttrName) override;
+};
+
+PXRUSDMAYA_REGISTER_SHADER_READER(UsdPreviewSurface, PxrMayaUsdPreviewSurface_Reader)
+
+PxrMayaUsdPreviewSurface_Reader::PxrMayaUsdPreviewSurface_Reader(
+    const UsdMayaPrimReaderArgs& readArgs)
+    : UsdMayaShaderReader(readArgs)
+{
+}
+
+/* virtual */
+bool PxrMayaUsdPreviewSurface_Reader::Read(UsdMayaPrimReaderContext* context)
+{
+    const auto&    prim = _GetArgs().GetUsdPrim();
+    UsdShadeShader shaderSchema = UsdShadeShader(prim);
+    if (!shaderSchema) {
+        return false;
+    }
+
+    MStatus           status;
+    MFnDependencyNode depFn;
+    if (!(UsdMayaTranslatorUtil::CreateShaderNode(
+              MString(prim.GetName().GetText()),
+              PxrMayaUsdPreviewSurfaceTokens->MayaTypeName.GetText(),
+              UsdMayaShadingNodeType::Shader,
+              &status,
+              &_mayaObject)
+          && depFn.setObject(_mayaObject))) {
+        // we need to make sure assumes those types are loaded..
+        TF_RUNTIME_ERROR(
+            "Could not create node of type '%s' for shader '%s'. "
+            "Probably missing a loadPlugin.\n",
+            PxrMayaUsdPreviewSurfaceTokens->MayaTypeName.GetText(),
+            prim.GetName().GetText());
+        return false;
+    }
+
+    MDGModifier modifier;
+    bool useModifier = false;
+    for (const UsdShadeInput& input : shaderSchema.GetInputs()) {
+        auto mayaAttrName = GetMayaNameForUsdAttrName(input.GetFullName());
+        if (mayaAttrName == TfToken()) {
+            continue;
+        }
+        MPlug mayaAttr = depFn.findPlug(mayaAttrName.GetText(), true, &status);
+        if (status == MS::kSuccess) {
+            if (mayaAttrName == PxrMayaUsdPreviewSurfaceTokens->DiffuseColorAttrName
+                || mayaAttrName == PxrMayaUsdPreviewSurfaceTokens->EmissiveColorAttrName
+                || mayaAttrName == PxrMayaUsdPreviewSurfaceTokens->SpecularColorAttrName) {
+                // These values are exported unscaled, so they also reimport unscaled:
+                VtValue val;
+                if (input.Get(&val) && val.IsHolding<GfVec3f>()) {
+                    GfVec3f color = val.UncheckedGet<GfVec3f>();
+                    // Use MDgModifier to skip color scaling done in SetMayaAttr
+                    MFnNumericData data;
+                    MObject        dataObj = data.create(MFnNumericData::k3Float);
+                    data.setData3Float(color[0], color[1], color[2]);
+                    modifier.newPlugValue(mayaAttr, dataObj);
+                    useModifier = true;
+                }
+            } else {
+                // This call always scales colors:
+                UsdMayaReadUtil::SetMayaAttr(mayaAttr, input);
+            }
+        }
+    }
+    if (useModifier) {
+        modifier.doIt();
```

Comment:
Why do we have to set it with DG modifier?

Reply:
Yes, I am actually curious why UsdMayaReadUtil::SetMayaAttr does that. It would seem easier to simply set the value directly into the MPlug.